src/extension.rs

```diff
@@ -0,0 +1,881 @@
+use crate::alloc::boxed::Box;
+use crate::alloc::collections::btree_map::Entry;
+use crate::alloc::collections::BTreeMap;
+use crate::alloc::fmt::{Display, Formatter};
+use crate::encoding::{DecodeContext, WireType};
+use crate::generic::{EncodeBuffer, Merge, MergeBuffer, ProtoIntType};
+use crate::{DecodeError, Encode};
+use bytes::{Buf, BufMut};
+use core::any::Any;
+use core::fmt::Debug;
+use core::marker::PhantomData;
+use core::option::Option;
+
+type ExtendableTypeId = &'static str;
+type FieldTag = u32;
+
+/// Marks a message as having able to have extensions.
+/// Extension data must be retrieved and modified through use of static Extensions in generated code.
+pub trait Extendable: 'static {
+    /// A static type id associated with this Extendable.
+    fn extendable_type_id() -> ExtendableTypeId;
+
+    /// The data structure that stores extension data.
+    fn extension_set(&self) -> &ExtensionSet<Self>;
+
+    /// The data structure that stores extension data.
+    fn extension_set_mut(&mut self) -> &mut ExtensionSet<Self>;
```

Comment:
These two methods basically require the implementation to have an `ExtensionSet` field, which feels unnecessarily restrictive. For example, another valid implementation would be to store unknown fields at decode time (without looking at `ExtensionRegistry` and then decode on-the-fly in methods like `extension_data`